bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/api/impl/EcovacsXmppDevice.java

```diff
@@ -0,0 +1,468 @@
+            throws EcovacsApiException {
+        EcovacsApiConfiguration config = api.getConfig();
+        PortalLoginResponse loginData = api.getLoginData();
+        if (loginData == null) {
+            throw new EcovacsApiException("Can not connect when not logged in");
+        }
+
+        logger.trace("{}: Connecting to XMPP", getSerialNumber());
+
+        String password = String.format("0/%s/%s", loginData.getResource(), loginData.getToken());
+        String host = String.format("msg-%s.%s", config.getContinent(), config.getRealm());
+
+        try {
+            Jid ownAddress = JidCreate.from(loginData.getUserId(), config.getRealm(), loginData.getResource());
+            Jid targetAddress = JidCreate.from(device.getDid(), device.getDeviceClass() + ".ecorobot.net", "atom");
+
+            XMPPTCPConnectionConfiguration connConfig = XMPPTCPConnectionConfiguration.builder().setHost(host)
+                    .setPort(5223).setUsernameAndPassword(loginData.getUserId(), password)
+                    .setResource(loginData.getResource()).setXmppDomain(config.getRealm())
+                    .setCustomX509TrustManager(TrustAllTrustManager.getInstance()).setSendPresence(false).build();
+
+            XMPPTCPConnection conn = new XMPPTCPConnection(connConfig);
+            conn.addConnectionListener(new ConnectionListener() {
+                @Override
+                public void connected(@Nullable XMPPConnection connection) {
+                }
+
+                @Override
+                public void authenticated(@Nullable XMPPConnection connection, boolean resumed) {
+                }
+
+                @Override
+                public void connectionClosed() {
+                }
+
+                @Override
+                public void connectionClosedOnError(@Nullable Exception e) {
+                    logger.trace("{}: XMPP connection failed", getSerialNumber(), e);
+                    if (e != null) {
+                        listener.onEventStreamFailure(EcovacsXmppDevice.this, e);
+                    }
+                }
+            });
+
+            PingHandler pingHandler = new PingHandler(conn, scheduler, listener, targetAddress);
+            messageHandler = new IncomingMessageHandler(listener);
+
+            Roster roster = Roster.getInstanceFor(conn);
+            roster.setRosterLoadedAtLogin(false);
+
+            conn.registerIQRequestHandler(messageHandler);
+            conn.connect();
+
+            this.connection = conn;
+            this.ownAddress = ownAddress;
+            this.targetAddress = targetAddress;
+            this.pingHandler = pingHandler;
+
+            conn.login();
+            conn.setReplyTimeout(1000);
+
+            logger.trace("{}: XMPP connection established", getSerialNumber());
+
+            listener.onFirmwareVersionChanged(this, sendCommand(new GetFirmwareVersionCommand()));
+            pingHandler.start();
+        } catch (EcovacsApiException e) {
```

Comment:
Only a small remark for consideration: I didn't check if you actually make use of the fact that `EcovacsApiException` is derived from `IOException`. In this case it works against you, because you need this additional catch/throw to avoid wrapping it into `EcovacsApiException` when already this type. If you mostly benefit from handling `EcovacsApiException` as `IOException`, this is fine, but otherwise you could consider making `EcovacsApiException` derive straight from `Exception` and remove this rethrow.

Reply:
The reason for deriving from IOException is that I wanted EcovacsApiException to be a checked exception, so that I don't / can't forget to handle it.

Reply:
Thanks, I wasn't aware of that :+1: Rethrow removed.